[lib]: STR Inline barrel shifters in AArch64.

lukeg101 opened the pull request:
Following https://github.com/herd/herdtools7/pull/31

Slight modification to tests following mixed-mode restriction on
allowable tests

It's what it says on the tin, unittests included.

lukeg101 commented:
Hi folks,
I have updated this patch with the following:
- Improved Commit title and description
- Added `Variant=mixed` to the unittests, we cannot run these tests just yet and we should wait to merge this as a result
- rebased onto most recent master
- added .expected files in line with the most recent infrastructure improvements.

I believe this code is good to go now, however we should not merge until we can run the unittests above.

Luc please could you take a look at `litmus/AArch64Compile_litmus.ml` - For `V32, RV`/`V64,RV` I have made sure to include `SXTW` if no shift is provided, but I cannot test this on an AArch64 machine right now. Please could you let me know if I need to add more tests?

lukeg101 commented:
Hi folks,
I'm just reopening this, @relokin has kindly reviewed this patch and I believe it is ready for public review! I have rebased it to the most recent herd, moved the mixed tests into the AArch64.mixed directory (non-mixed test available too)
Please let me know what you think
Thanks
Luke

maranget commented:
Hi @lukeg101 Looks pretty clean. Some tests of actual usage of barrel shifter would be a real bonus. If I am not wrong the tests are  simple constant and register offsets. Or did I miss something?

maranget commented on herd/tests/instructions/AArch64.mixed/A14.litmus:
```diff
@@ -0,0 +1,9 @@
+AArch64 A14
+(* Store register, with offset, symbolic location *)
+
+ { uint64_t x; 0:x0=1; 0:x2=x; 0:x3=4;}
```
Declaring register 0:x3 to be used as a 64bits register as follows would avoid a (litmus) warnig.
```suggestion
 { uint64_t x; 0:x0=1; 0:x2=x; int64_t 0:x3=4;}
```

lukeg101 commented:
> Hi @lukeg101 Looks pretty clean. Some tests of actual usage of barrel shifter would be a real bonus. If I am not wrong the tests are simple constant and register offsets. Or did I miss something?

Good spot thanks! When I initially wrote this patch we needed something to make sense of an address offset like `v+8` now we have a vector type I think it is possible to test the barrel shifters. I will see if I can conjure up some tests!

maranget commented:
Some tests use arrrays already. For instance M000 and M0001 in AArch64.mixed tests (there are others in non-mixed-size tests). If you have access to a ARMv8 machine, it is a good idea to check the tests using litmus (I have just corrected M001 in master thanks to this reality check).

lukeg101 commented:
Thanks that's good to know for the sake of formatting! I will put some tests together and test litmus on a TX2 before submitting 👍

relokin commented:
It would be great to add tests with vectors too but something like this would also test shifts:
```
AArch64 A
{ uint64_t x; 0:x0=1; 0:x2=x; int64_t 0:x3=1;}
P0                                       ;
   STR W0, [X2, X3, LSL#2];
exists (x=4294967296)
```

maranget commented:
Tests such as yours are  appropriate. The advantage of vectors is the possibility to write tests that do not require the mixed-size option (`-variant mixed`). See [test L013](https://github.com/herd/herdtools7/blob/master/herd/tests/instructions/AArch64/L013.litmus) for instance.

maranget commented:
Oups closed by mistake, sorry for this.

lukeg101 commented:
Hi folks,
Thank you for your feedback and discussion, hopefully I have addressed your comments above with:
- Adding Nikos' test `M002.litmus`
- Added non-vector mixed tests `M003`, `M004`, `M005`
- Added non-vector non-mixed tests `L015`, `L016`
- Added vector non-mixed test `L017`
- Added vector mixed test `M006`
- Tweak the type information in init so no warnings occur in litmus
- Tested on Thunder X2 (log below)

```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/A12.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 A12

{int64_t x=0; 0:X0=1; 0:X2=x;}

 P0             ;
 STR W0,[X2,#4] ;

exists (x=4294967296)
Generated assembler
#START _litmus_P0
	str w3,[x1,#4]

Test A12 Allowed
Histogram (1 states)
1000000*>x=4294967296;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=4294967296) is validated
Hash=86545e2af1a0880e99658d8d4d00c83c
Observation A12 Always 1000000 0
Time A12 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/A14.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 A14

{uint64_t x=0; 0:X0=1; 0:X2=x; int64_t 0:X3=4;}

 P0             ;
 STR W0,[X2,X3] ;

exists (x=4294967296)
Generated assembler
#START _litmus_P0
	str w3,[x1,x6]

Test A14 Allowed
Histogram (1 states)
1000000*>x=4294967296;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=4294967296) is validated
Hash=020d200946f57025e6078483a217d42c
Observation A14 Always 1000000 0
Time A14 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/M002.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 M002

{uint64_t x=0; 0:X0=1; 0:X2=x; int64_t 0:X3=1;}

 P0                    ;
 STR W0,[X2,X3,LSL #2] ;

exists (x=4294967296)
Generated assembler
#START _litmus_P0
	str w3,[x1,x3,LSL #2]

Test M002 Allowed
Histogram (1 states)
1000000*>x=4294967296;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=4294967296) is validated
Hash=822a4c2b0005fec10225b46de30bd2fe
Observation M002 Always 1000000 0
Time M002 0.04

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/M003.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 M003

{uint64_t x=0; 0:X0=1; 0:X2=x; int64_t 0:X3=0;}

 P0                    ;
 STR W0,[X2,X3,LSL #0] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str w3,[x1,x6,LSL #0]

Test M003 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=d612e42cf7c0844663af34d30297f2f6
Observation M003 Always 1000000 0
Time M003 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/M004.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 M004

{uint64_t x=0; 0:X0=1; 0:X2=x; int32_t 0:X3=0;}

 P0                  ;
 STR W0,[X2,W3,SXTW] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str w3,[x1,w6,SXTW]

Test M004 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=1e61a83b6f7566c8526ca4540c009b2f
Observation M004 Always 1000000 0
Time M004 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/M005.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 M005

{uint64_t x=0; 0:X0=1; 0:X2=x; int32_t 0:X3=0;}

 P0                  ;
 STR W0,[X2,W3,UXTW] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str w3,[x1,w6,UXTW]

Test M005 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=2c1dc0ae671fb6f35401c54bde5b7845
Observation M005 Always 1000000 0
Time M005 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64.mixed/M006.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 M006

{uint32_t x[2]; 0:X0=1; 0:X2=x; int64_t 0:X3=1;}

 P0                    ;
 STR W0,[X2,X3,LSL #2] ;

exists (x={2,1})
Generated assembler
#START _litmus_P0
	str w3,[x1,x3,LSL #2]

Test M006 Allowed
Histogram (1 states)
1000000*>x={2,1};
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x={2,1}) is validated
Hash=4d6e36211c150aee5ee2723464a442f6
Observation M006 Always 1000000 0
Time M006 0.08

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64/A10.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 A10

{int64_t x=0; int64_t 0:X0=1; 0:X2=x;}

 P0          ;
 STR X0,[X2] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str x3,[x1]

Test A10 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=ea5191f717adc74412c7fe1dae4db891
Observation A10 Always 1000000 0
Time A10 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64/L015.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 L015

{uint64_t x=0; 0:X0=1; 0:X2=x; int32_t 0:X3=0;}

 P0                  ;
 STR W0,[X2,W3,SXTW] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str w3,[x1,w6,SXTW]

Test L015 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=1e61a83b6f7566c8526ca4540c009b2f
Observation L015 Always 1000000 0
Time L015 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64/L016.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 L016

{uint64_t x=0; 0:X0=1; 0:X2=x; int32_t 0:X3=0;}

 P0                  ;
 STR W0,[X2,W3,UXTW] ;

exists (x=1)
Generated assembler
#START _litmus_P0
	str w3,[x1,w6,UXTW]

Test L016 Allowed
Histogram (1 states)
1000000*>x=1;
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x=1) is validated
Hash=2c1dc0ae671fb6f35401c54bde5b7845
Observation L016 Always 1000000 0
Time L016 0.05

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/AArch64/L017.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
AArch64 L017

{uint32_t x[2]; 0:X0=1; 0:X2=x; int64_t 0:X3=1;}

 P0                    ;
 STR W0,[X2,X3,LSL #2] ;

exists (x={2,1})
Generated assembler
#START _litmus_P0
	str w3,[x1,x3,LSL #2]

Test L017 Allowed
Histogram (1 states)
1000000*>x={2,1};
Ok

Witnesses
Positive: 1000000, Negative: 0
Condition exists (x={2,1}) is validated
Hash=4d6e36211c150aee5ee2723464a442f6
Observation L017 Always 1000000 0
Time L017 0.07

Machine: <CPU INFORMATION OMITTED BY LUKE - NOT RELEVANT>

Revision c87d595dc6355ce15b57bcef7b2231a1466c0c65, version 7.56+02~dev
Command line: ./_build/install/default/bin/litmus7 herd/tests/instructions/AArch64.mixed/A12.litmus herd/tests/instructions/AArch64.mixed/A14.litmus herd/tests/instructions/AArch64.mixed/M002.litmus herd/tests/instructions/AArch64.mixed/M003.litmus herd/tests/instructions/AArch64.mixed/M004.litmus herd/tests/instructions/AArch64.mixed/M005.litmus herd/tests/instructions/AArch64.mixed/M006.litmus herd/tests/instructions/AArch64/A10.litmus herd/tests/instructions/AArch64/L015.litmus herd/tests/instructions/AArch64/L016.litmus herd/tests/instructions/AArch64/L017.litmus -set-libdir litmus/libdir/
Parameters
#define SIZE_OF_TEST 100000
#define NUMBER_OF_RUN 10
#define AVAIL 1
#define STRIDE (-1)
#define MAX_LOOP 0
/* gcc options: -Wall -std=gnu99 -O2 -pthread */
/* barrier: user */
/* launch: changing */
/* affinity: none */
/* alloc: dynamic */
/* memory: direct */
/* safer: write */
/* preload: random */
/* speedcheck: no */
```

relokin commented on herd/tests/instructions/AArch64.mixed/M006.litmus:
```diff
@@ -0,0 +1,10 @@
+AArch64 M006
```
Do we need mixed mode for this test?

lukeg101 replied:
After a quick run in Herd, no we don't - will move to non-mixed directory

lukeg101 commented:
Instead of moving `M006` to the non-mixed test directory, I copied it to `L018` as there is no harm in doubling the tests. I did not re-run litmus as it's a duplicate test however
